### PR TITLE
Patch grover

### PR DESCRIPTION
**Pull request by Mayank447**
- Merges master in gate_refactor
- Patches the Grover search implementation #314 in accordances with gate refactor. Check commit [Patch grover algorithm in accordance with gate refactor](https://github.com/qutip/qutip-qip/pull/346/commits/50b87924f7d34e3ac6787c124f031b248236910e) for details

### CODE REVIEW

File: src/qutip_qip/algorithms/grover.py [hunk 1]
```diff
@@ -187,7 +182,13 @@ def grover(
...
         # Oracle
         if isinstance(oracle, QubitCircuit):
-            qc.gates.extend(oracle.gates)
+            for circ_inst in oracle.instructions:
+                qc.add_gate(
+                    circ_inst.operation,
+                    targets=circ_inst.targets,
+                    controls=circ_inst.controls,
+                )
+
```
Comment (BoxiLi) [inline src/qutip_qip/algorithms/grover.py:191]: We might want an `append_circuit` function at some time. Maybe it is in your proposal of Ops.

Comment (Mayank447) [reply src/qutip_qip/algorithms/grover.py:191]: Yes, if oracle was an `Ops`. `qc.add_op(oracle, target=qubits)` will do the above.

**Comment by Mayank447** [comment]
@BoxiLi The suggestion as per Copilot are unrelated to the patch. So I am gonna open a separate issue for these. These should be addressed in a separate PR.

**Comment by Mayank447** [comment]
You can go ahead with the merge if it looks good to you.

File: tests/test_grover.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        oracle = grover_oracle(2, 3)
+
+        with pytest.raises(
+            ValueError, match="num_iterations must not be a negative"
+        ):
+            grover(oracle, 2, 1, num_iterations=-3)
+
+    def test_grover_1_qubit(self):
+        """
+        Full algoritm test: 1 qubit(edge-case), searching for |0>.
```
[the remark this answers is not in the document]
Comment (Mayank447) [reply tests/test_grover.py:56]: This is a redundant change

**Comment by BoxiLi** [comment]
Sure, make sense, so we merge this as is?

File: src/qutip_qip/algorithms/grover.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    # Grover Iterations:
+    for _ in range(num_iterations):
+        # Oracle
+        if isinstance(oracle, QubitCircuit):
+            for circ_inst in oracle.instructions:
+                qc.add_gate(
+                    circ_inst.operation,
+                    targets=circ_inst.targets,
+                    controls=circ_inst.controls,
```
[the remark this answers is not in the document]
Comment (Mayank447) [reply src/qutip_qip/algorithms/grover.py:189]: It should be classical_controls, classical_controls_value instead of cbits, cbits_ctrl_value.
